src/clib/pioc.c

```diff
@@ -1451,6 +1451,430 @@ int PIOc_Init_Async(MPI_Comm world, int num_io_procs, int *io_proc_list,
+        return check_mpi(NULL, ret, __FILE__, __LINE__);
+    LOG((3, "world group created\n"));
+
+    /* We will create a group for the IO component. */
+    MPI_Group io_group;
+
+    /* The shared IO communicator. */
+    MPI_Comm io_comm;
+
+    /* Rank of current process in IO communicator. */
+    int io_rank = -1;
+
+    /* Set to MPI_ROOT on master process, MPI_PROC_NULL on other
+     * processes. */
+    int iomaster;
+
+    /* Create a group for the IO component. */
+    if ((ret = MPI_Group_incl(world_group, num_io_procs, my_io_proc_list, &io_group)))
+        return check_mpi(NULL, ret, __FILE__, __LINE__);
+    LOG((3, "created IO group - io_group = %d MPI_GROUP_EMPTY = %d", io_group, MPI_GROUP_EMPTY));
+
+    /* There is one shared IO comm. Create it. */
+    if ((ret = MPI_Comm_create(world, io_group, &io_comm)))
+        return check_mpi(NULL, ret, __FILE__, __LINE__);
+    LOG((3, "created io comm io_comm = %d", io_comm));
+
+    /* Does the user want a copy of the IO communicator? */
+    if (user_io_comm)
+    {
+        *user_io_comm = MPI_COMM_NULL;
+        if (in_io)
+            if ((mpierr = MPI_Comm_dup(io_comm, user_io_comm)))
+                return check_mpi(NULL, mpierr, __FILE__, __LINE__);
+    }
+
+    /* For processes in the IO component, get their rank within the IO
+     * communicator. */
+    if (in_io)
+    {
+        LOG((3, "about to get io rank"));
+        if ((ret = MPI_Comm_rank(io_comm, &io_rank)))
+            return check_mpi(NULL, ret, __FILE__, __LINE__);
+        iomaster = !io_rank ? MPI_ROOT : MPI_PROC_NULL;
+        LOG((3, "intracomm created for io_comm = %d io_rank = %d IO %s",
+             io_comm, io_rank, iomaster == MPI_ROOT ? "MASTER" : "SERVANT"));
+    }
+
+    /* We will create a group for each computational component. */
+    MPI_Group group[component_count];
+
+    /* We will also create a group for each component and the IO
+     * component processes (i.e. a union of computation and IO
+     * processes. */
+    MPI_Group union_group[component_count];
+
+    /* For each computation component. */
+    for (int cmp = 0; cmp < component_count; cmp++)
+    {
+        LOG((3, "processing component %d", cmp));
+
+        /* Get pointer to current iosys. */
+        my_iosys = iosys[cmp];
+        
+        /* Initialize some values. */
+        my_iosys->io_comm = MPI_COMM_NULL;
+        my_iosys->comp_comm = MPI_COMM_NULL;
+        my_iosys->union_comm = MPI_COMM_NULL;
+        my_iosys->intercomm = MPI_COMM_NULL;
+        my_iosys->my_comm = MPI_COMM_NULL;
+        my_iosys->async_interface = 1;
+        my_iosys->error_handler = default_error_handler;
+        my_iosys->num_comptasks = num_procs_per_comp[cmp];
+        my_iosys->num_iotasks = num_io_procs;
+        my_iosys->compgroup = MPI_GROUP_NULL;
+        my_iosys->iogroup = MPI_GROUP_NULL;
+        
+        /* The rank of the computation leader in the union comm. */
+        my_iosys->comproot = num_io_procs;
```

Comment:
OK, noted. Let me ponder that. However I will still move forward with this PR because it is an improvement on the current code.